tools/logger/filter.c

```diff
@@ -0,0 +1,330 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
+ *
+ * Author: Karol Trzcinski	<karolx.trzcinski@linux.intel.com>
+ */
+
+#include <ipc/trace.h>
+#include <smex/ldc.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <user/trace.h>
+#include <ctype.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "convert.h"
+#include "filter.h"
+
+#define COMPONENTS_SEPARATOR ','
+
+/** map between log level given by user and enum value */
+static const struct {
+	const char name[16];
+	int32_t log_level;
+} log_level_dict[] = {
+	{"verbose",	LOG_LEVEL_VERBOSE},
+	{"debug",	LOG_LEVEL_DEBUG},
+	{"info",	LOG_LEVEL_INFO},
+	{"warning",	LOG_LEVEL_WARNING},
+	{"error",	LOG_LEVEL_ERROR},
+	{"critical",	LOG_LEVEL_CRITICAL},
+	{"v",		LOG_LEVEL_VERBOSE},
+	{"d",		LOG_LEVEL_DEBUG},
+	{"i",		LOG_LEVEL_INFO},
+	{"w",		LOG_LEVEL_WARNING},
+	{"e",		LOG_LEVEL_ERROR},
+	{"c",		LOG_LEVEL_CRITICAL},
+};
+
+struct filter_element {
+	struct list_item list;
+	int32_t uuid_id;	/**< type id, or -1 when not important */
+	int32_t comp_id;	/**< component id or -1 when not important */
+	int32_t pipe_id;	/**< pipeline id or -1 when not important */
+	int32_t log_level;	/**< new log level value */
+};
+
+static char *ltrim(char *s)
+{
+	while (isspace(*s))
+		s++;
+	return s;
+}
+
+static char *rtrim(char *s)
+{
+	char *ptr = s + strlen(s) - 1;
+
+	while (ptr >= s && isspace(*ptr)) {
+		*ptr = '\0';
+		--ptr;
+	}
+	return s;
+}
+
+static char *trim(char *s)
+{
+	return ltrim(rtrim(s));
+}
+
+/**
+ * Search for uuid entry with given component name in given uids dictionary
+ * @param uids_dict dictionary to search in
+ * @param name of uuid entry
+ * @return pointer to sof_uuid_entry with given name
+ */
+static struct sof_uuid_entry *
+get_uuid_by_name(const struct snd_sof_uids_header *uids_dict,
+		 const char *name)
+{
+	uintptr_t beg = (uintptr_t)uids_dict + uids_dict->data_offset;
+	uintptr_t end = beg + uids_dict->data_length;
+	struct sof_uuid_entry *ptr = (struct sof_uuid_entry *)beg;
+
+	while ((uintptr_t)ptr < end) {
+		if (strcmp(name, ptr->name) == 0)
+			return ptr;
+		++ptr;
+	}
+	return NULL;
+}
+
+/**
+ * Parse log level name (from log_level_dict) to enum value.
+ * Take care about possible whitespace at the begin.
+ * @param value_start pointer to the begin of range to search
+ * @return enum value for given log level, or -1 for invalid value
+ */
+static int filter_parse_log_level(const char *value_start)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(log_level_dict); ++i) {
+		if (strstr(log_level_dict[i].name, value_start))
+			return log_level_dict[i].log_level;
+	}
+	return -1;
+}
+
+static char *filter_parse_component_name(const struct snd_sof_uids_header *uids_dict,
+					 char *input_str,
+					 struct filter_element *out)
+{
+	static char scan_format_string[32] = "";
```

Comment:
Is 32 bytes for a single component or command line ? This should also be a macro.